model/tetra/tetra_material_4d.cpp

```diff
@@ -122,6 +123,52 @@ void TetraMaterial4D::set_albedo_source(const TetraColorSource p_albedo_source)
 	_albedo_source = p_albedo_source;
 	_albedo_source_flags = _tetra_source_to_flags(_albedo_source);
 	notify_property_list_changed();
+	update_cross_section_material();
+}
+
+Ref<Texture3D> TetraMaterial4D::get_texture() const {
+	return _texture;
+}
+
+void TetraMaterial4D::set_texture(const Ref<Texture3D> &p_texture) {
+	_texture = p_texture;
+	update_cross_section_material();
+}
+
+void TetraMaterial4D::update_cross_section_material() {
+	if (_cross_section_material.is_null()) {
+		return;
+	}
+	if (_cross_section_material->get_shader().is_null()) {
+		// TODO this re-compiles the shader for every material, should cache the Shader object somewhere.
+		Ref<Shader> cross_section_shader;
+		cross_section_shader.instantiate();
+		cross_section_shader->set_code(cross_section_shader_shader_glsl);
+		_cross_section_material->set_shader(cross_section_shader);
+	}
+	Color albedo;
+	Variant texture;
+	switch (_albedo_source) {
+		case TETRA_COLOR_SOURCE_SINGLE_COLOR:
+			albedo = _albedo_color;
+			// Setting to a Nil variant resets to the default texture, which is white.
+			texture = Variant();
+			break;
+		case TETRA_COLOR_SOURCE_CELL_UVW_ONLY:
+			albedo = Color(1.0, 1.0, 1.0);
+			texture = _texture;
+			break;
+		case TETRA_COLOR_SOURCE_CELL_UVW_AND_SINGLE:
+			albedo = _albedo_color;
+			texture = _texture;
+			break;
+		default:
+			albedo = Color(1.0, 1.0, 1.0);
+			texture = Variant();
+			break;
+	}
```

Comment:
The TetraColorSource enum is intended for safe user-friendly setting, but when reading, there are bitwise flags available to simplify this. That code can be replaced with this:

```cpp
	const ColorSourceFlags flags = get_albedo_source_flags();
	const Color albedo = (flags & Material4D::COLOR_SOURCE_FLAG_SINGLE_COLOR) ? _albedo_color : Color(1.0, 1.0, 1.0);
	// Setting to a Nil variant resets to the default texture, which is white.
	const Variant texture = (flags & Material4D::COLOR_SOURCE_FLAG_CELL_UVW) ? Variant(_texture) : Variant();
```

(feel free to put this in your new PR)